File: StkExtensionPlugins/PluginScripts/Matlab/ReceiverModel.m
function [output] = Matlab_ReceiverModel(input)
	% NOTE: the outputs that are returned
	%       MUST be in the same order as registered

	% NOTE: Please DO NOT change anything above or below the USER MODEL AREA.
	% 
	% This example shows static output parameters, but these parameters
      % may change value at each time step.

switch input.method
    
    case 'register'
 	% register output variables       
        
	% The frequency of the receiver is a double in Hz.
		Frequency= {'ArgumentName','Frequency',...
                     'Name','Frequency',...
                     'ArgumentType','Output'};
    
	% The bandwidth of the RF spectrum is a double in Hz.
	   Bandwidth= {'ArgumentName','Bandwidth',...
                    'Name','Bandwidth',...
                    'ArgumentType','Output'};

	% The gain of the radiating elements of the antenna is a double in dB or dBi.
        Gain= {'ArgumentName','Gain',...
                     'Name','Gain',...
                     'ArgumentType','Output'};
    
	% A collection of pre-receive losses is a double in dB.
        PreReceiveLoss= {'ArgumentName','PreReceiveLoss',...
                    'Name','PreReceiveLoss',...
                    'ArgumentType','Output'};

	% A collection of pre-demodulation losses is a double in dB.
	   PreDemodLoss= {'ArgumentName','PreDemodLoss',...
                     'Name','PreDemodLoss',...
                     'ArgumentType','Output'};
    
	% UseRainModel is a flag (0 or 1) that specifies whether a rain model is to be used which is Boolean.
        UseRainModel= {'ArgumentName','UseRainModel',...
                    'Name','UseRainModel',...
                    'ArgumentType','Output'};
					
	% The percent outage value that can be tolerated by the link is a double.
        RainOutagePercent= {'ArgumentName','RainOutagePercent',...
                     'Name','RainOutagePercent',...
                     'ArgumentType','Output'};
    
	% The type of polarization is an Integer:
	% Value 	Polarization Type 		Required Parameter(s)
	%	0 		None 					None
	%	1 		Linear 					Reference Axis
	%	2		Right Hand Circular 	None
	% 	3		Left Hand Circular 		None
	%	4 		Vertical 				Reference Axis, Tilt Angle
	%	5 		Horizontal 				Reference Axis, Tilt Angle
	%	6 		Elliptical 				Reference Axis, Tilt Angle, Axial Ratio
	
        PolType= {'ArgumentName','PolType',...
                    'Name','PolType',...
                    'ArgumentType','Output'};

	%	The polarization reference axis is an Integer that is used to align transmitter polarization to receiver polarization, with 	
	%	0, 1 and 2 representing the X, Y and Z axes, respectively.       
	   PolRefAxis= {'ArgumentName','PolRefAxis',...
                     'Name','PolRefAxis',...
                     'ArgumentType','Output'};
    
	%  The polarization tilt angle measured from the reference axis is a double in degrees.
        PolTiltAngle= {'ArgumentName','PolTiltAngle',...
                    'Name','PolTiltAngle',...
                    'ArgumentType','Output'};

    %  The polarization axial ratio is a real.  
		PolAxialRatio= {'ArgumentName','PolAxialRatio',...
                     'Name','PolAxialRatio',...
                     'ArgumentType','Output'};
					 
	% The Cross Polarization Leakage is a double in dB ranging from -9999.9 dB to 0 dB where -9999.9 represents the ideal case
	% of no cross polarization leakage.  
		CrossPolLeakage = {'ArgumentName','CrossPolLeakage',...
                    'Name','CrossPolLeakage',...
                    'ArgumentType','Output'};
					
	% CalcReceiverNoiseTemp is a flag (0 or 1) that specifies whether the Receiver Noise Temperature is to be calculated or a 
	% constant which is Boolean. 0 indicates a constant system noise temperature and 1 indicates calculated system noise 
	% temperature.
        CalcReceiverNoiseTemp = {'ArgumentName','CalcReceiverNoiseTemp',...
                    'Name','CalcReceiverNoiseTemp',...
                    'ArgumentType','Output'};
					
	% The constant system noise temperature (when CalcReceiverNoiseTemp = 0) is a double in K.
        ConstantNoiseTemp = {'ArgumentName','ConstantNoiseTemp',...
                    'Name','ConstantNoiseTemp',...
                    'ArgumentType','Output'};
    
    % The receiver pre-amplifier noise figure is a double in dB.  
		ReceiverNoiseFigure= {'ArgumentName','ReceiverNoiseFigure',...
                    'Name','ReceiverNoiseFigure',...
                    'ArgumentType','Output'};

	% The cable loss value is a double in dB.
        CableLoss= {'ArgumentName','CableLoss',...
                     'Name','CableLoss',...
                     'ArgumentType','Output'};
					 
    % The temperature of the cable is a double in K.
        CableNoiseTemp= {'ArgumentName','CableNoiseTemp',...
                    'Name','CableNoiseTemp',...
                    'ArgumentType','Output'};
    
	%	The receiver antenna noise temperature is a double in K.
        AntennaNoiseTemp= {'ArgumentName','AntennaNoiseTemp',...
                    'Name','AntennaNoiseTemp',...
                    'ArgumentType','Output'};

 	% register input variables       
   
	% The current date and time is a String variable. 
		date = {'ArgumentName','DateUTC',...
                'Name','DateUTC',...
                'ArgumentType','Input'};
 
    % The scenario central body is a String variable. 
		CbName = {'ArgumentName','CbName',...
                  'Name','CbName',...
                  'ArgumentType','Input'};
				  
	% The scenario simulation epoch time is a double in seconds.
	   EpochSec = {'ArgumentName','EpochSec',...
                'Name','EpochSec',...
                'ArgumentType','Input'};

    % The transmitter position in Central Body Fixed coordinates is a vector of doubles, of length 3, corresponding to the X, Y and Z
	% values with the unit of meters.
		XmtrPosCBF = {'ArgumentName','XmtrPosCBF',...
               'Name','XmtrPosCBF',...
               'ArgumentType','Input'};

	% The attitude quaternion of the transmitter is a vector of doubles, of length 4.
        XmtrAttitude= {'ArgumentName','XmtrAttitude',...
               'Name','XmtrAttitude',...
               'ArgumentType','Input'};
    
	% The receiver position in Central Body Fixed coordinates is a vector of doubles, of length 3, corresponding to the X, Y and Z
	% values with the unit of meters.
        RcvrPosCBF= {'ArgumentName','RcvrPosCBF',...
               'Name','RcvrPosCBF',...
               'ArgumentType','Input'};
 
	% The attitude quaternion of the receiver is a vector of doubles, of length 4.
        RcvrAttitude= {'ArgumentName','RcvrAttitude',...
               'Name','RcvrAttitude',...
               'ArgumentType','Input'};
   
        output = {Frequency, Bandwidth, Gain, PreReceiveLoss, PreDemodLoss,...
                  UseRainModel, RainOutagePercent,...
                  PolType, PolRefAxis, PolTiltAngle, PolAxialRatio,...
				  CrossPolLeakage, CalcReceiverNoiseTemp, ConstantNoiseTemp,...
                  ReceiverNoiseFigure, CableLoss, CableNoiseTemp, AntennaNoiseTemp,...
                  date, CbName, EpochSec, XmtrPosCBF, XmtrAttitude, RcvrPosCBF, RcvrAttitude};
    
    case 'compute'
    
        computeData = input.methodData;

	% compute the Test Model : 
	% Example Model for testing only
	% Receiver input & output parameter usage is shown
	% 

	% USER Receiver MODEL AREA.
        

	  xmtrPos = [ 0 0 0 ];
	  rcvrPos = [ 0 0 0 ];
        xmAttQuat = [ 0 0 0 0 ];
        rcAttQuat = [ 0 0 0 0 ];
	  xmtrPos = computeData.XmtrPosCBF;
	  rcvrPos = computeData.RcvrPosCBF;
        xmAttQuat = computeData.XmtrAttitude;
        rcAttQuat = computeData.RcvrAttitude;

	  range = norm(xmtrPos  - rcvrPos);

        output.Frequency= 12.3e9;
        output.Bandwidth= 25.0e6;
        output.Gain= 30;
        output.PreReceiveLoss= -2.0;
        output.PreDemodLoss= 0.0;
        output.UseRainModel= 0;
        output.RainOutagePercent= 0.3;
        output.PolType= 0;
        output.PolRefAxis= 0;
        output.PolTiltAngle= 0.0;
        output.PolAxialRatio= 1.0;
		output.CrossPolLeakage= -60.0;
		output.CalcReceiverNoiseTemp = 0;
		output.ConstantNoiseTemp = 290;
        output.ReceiverNoiseFigure= 0.25;
        output.CableLoss= 0.5;
        output.CableNoiseTemp= 290;
        output.AntennaNoiseTemp= 50.0;

	% END OF USER MODEL AREA

    otherwise
        output = [];
end

File: StkExtensionPlugins/PluginScripts/Perl/ReceiverModel.pl
##########################################################################################
# SAMPLE FOR PERL BASED CUSTOM TRANSMITTER MODEL PLUGIN SCRIPT PROVIDED BY THE USER
# PLEASE ADD YOUR MODEL IN THE USER ANTENNA GAIN MODEL AREA BELOW.
# DO NOT CHANGE ANYTHING ELSE IN THE SCRIPT
##########################################################################################

# Perl_ReceiverModel

# declare some global variablesfunction

my $Perl_ReceiverModel_compute_init = -1;
my $Perl_ReceiverModel_Inputs, $Perl_ReceiverModel_Outputs;
my %Perl_ReceiverModel_Outputs_ArgHash;

sub Perl_ReceiverModel
{
# the inputs to the the script arise as a reference to an array
	# the STKUtil::getInputArray function is used to get at the array itself

	my @inputData = @{$_[0]};
 
	my @retVal;

	if ( !defined($inputData[0]) )
	{
		# do compute

		@retVal = Perl_ReceiverModel_compute(@inputData);

	}
	elsif ( $inputData[0] eq 'register' )
	{
		$Perl_ReceiverModel_compute_init = -1;

		@retVal = Perl_ReceiverModel_register();

      }
	elsif ( $inputData[0] eq 'compute' )
	{
		@retVal = Perl_ReceiverModel_compute(@inputData);
      }
	else
	{
		# error: do nothing
	}

	# MUST return a reference to an array, as shown below

	return \@retVal;
}

sub Perl_ReceiverModel_register
{
	my @argStr;

	push @argStr, "ArgumentType = Output; Name = Frequency; ArgumentName = Frequency";
	push @argStr, "ArgumentType = Output; Name = Bandwidth; ArgumentName = Bandwidth";
	push @argStr, "ArgumentType = Output; Name = Gain; ArgumentName = Gain";
	push @argStr, "ArgumentType = Output; Name = PreReceiveLoss; ArgumentName = PreReceiveLoss";
	push @argStr, "ArgumentType = Output; Name = PreDemodLoss; ArgumentName = PreDemodLoss";
	push @argStr, "ArgumentType = Output; Name = UseRainModel; ArgumentName = UseRainModel";
	push @argStr, "ArgumentType = Output; Name = RainOutagePercent; ArgumentName = RainOutagePercent";
	push @argStr, "ArgumentType = Output; Name = PolType; ArgumentName = PolType";
	push @argStr, "ArgumentType = Output; Name = PolRefAxis; ArgumentName = PolRefAxis";
	push @argStr, "ArgumentType = Output; Name = PolTiltAngle; ArgumentName = PolTiltAngle";
	push @argStr, "ArgumentType = Output; Name = PolAxialRatio; ArgumentName = PolAxialRatio";
	push @argStr, "ArgumentType = Output; Name = CrossPolLeakage; ArgumentName = CrossPolLeakage";
	push @argStr, "ArgumentType = Output; Name = ReceiverNoiseFigure; ArgumentName = ReceiverNoiseFigure";
	push @argStr, "ArgumentType = Output; Name = CableLoss; ArgumentName = CableLoss";
	push @argStr, "ArgumentType = Output; Name = CableNoiseTemp; ArgumentName = CableNoiseTemp";
	push @argStr, "ArgumentType = Output; Name = AntennaNoiseTemp; ArgumentName = AntennaNoiseTemp";
	
	push @argStr, "ArgumentType = Input;  Name = DateUTC;           ArgumentName = DateUTC";
	push @argStr, "ArgumentType = Input;  Name = CbName;            ArgumentName = CbName";
	push @argStr, "ArgumentType = Input;  Name = XmtrPosCBF;        ArgumentName = XmtrPosCBF";
	push @argStr, "ArgumentType = Input;  Name = XmtrAttitude;      ArgumentName = XmtrAttitude";
	push @argStr, "ArgumentType = Input;  Name = RcvrPosCBF;      	ArgumentName = RcvrPosCBF";
	push @argStr, "ArgumentType = Input;  Name = RcvrAttitude;      ArgumentName = RcvrAttitude";
	push @argStr, "ArgumentType = Input;  Name = EpochSec;          ArgumentName = EpochSec";
	
	return @argStr; 
}


sub Perl_ReceiverModel_compute
{
	# the inputs here are in the order of the requested Inputs, as registered
	my @origArray = @_;

	# $origArray[0] is the calling mode

	# next argument is Date and is a String. Strings are simply scalars in Perl so the assignment is easy:	
	my $date = $origArray[1];

	# next argument is CbName and is a String. 	
	my $CbName = $origArray[2];
	
	# next argument is named XmtrPosCBF and is type (CartVec3) Array and is double-3. 
		
	my @XmtrPosCBF = @{$origArray[3]};
		
	# next argument is named XmtrAttitude and is type (Quaternion) Array and is double-4. 
		
	my @XmtrAttitude = @{$origArray[4]};
	
	# next argument is named RcvrPosCBF and is type (CartVec3) Array and is double-3. 
	
	my @RcvrPosCBF = @{$origArray[5]};
	
	# next argument is named RcvrAttitude and is type (Quaternion) Array and is double-4. 
	
	my @RcvrAttitude = @{$origArray[6]};
	
	# next argument is Time and is a value. Values are simply scalars in Perl so the assignment is easy:	
	my $time = $origArray[7];
	
	if($Perl_ReceiverModel_compute_init < 0)
	{
		$Perl_ReceiverModel_compute_init = 1; 

		# The following hashes have been created automatically after this script has registered its inputs and outputs.
		# Each hash contains information about the arguments for this script. The hashes have been created as a
		# user convenience, for those users wanting to know, during the running of the script, what the inputs
		# and outputs are. In many cases, the script write doesn't care, in which case this entire if-block
		# is unneeded and can be removed.

		$Perl_ReceiverModel_Inputs = $g_PluginArrayInterfaceHash{'Perl_ReceiverModel'}{'Inputs'};
		$Perl_ReceiverModel_Outputs = $g_PluginArrayInterfaceHash{'Perl_ReceiverModel'}{'Outputs'};

		%Perl_ReceiverModel_Outputs_ArgHash = $Perl_ReceiverModel_Outputs->getArgumentHash();

		# comment out the line below if you don't want to see the inputs and outputs each time the script is run
		Perl_ReceiverModel_showArgs();
	}

	# NOTE: the outputs that are returned MUST be in the same order as registered
	
	#############################################################################################
	# USER TRANSMITTER MODEL AREA.
	# PLEASE REPLACE THE CODE BELOW WITH YOUR TRANSMITTER MODEL
	#############################################################################################

	# compute the Test Model : Receiver Model
	
	my $fromX, $fromY, $fromZ, $toX, $toY, $toZ, $rX, $rY, $rZ;
	my $xmA, $xmB, $xmC, $xmD, $rcA, $rcB, $rcC, $rcD, $range;

	$fromX = @XmtrPosCBF[0];
	$fromY = @XmtrPosCBF[1];
	$fromZ = @XmtrPosCBF[2];
	
	$toX = @RcvrPosCBF[0];
	$toY = @RcvrPosCBF[1];
	$toZ = @RcvrPosCBF[2];

	$rX = $fromX - $toX;
	$rY = $fromY - $toY;
	$rZ = $fromZ - $toZ;
	
	$xmA = @XmtrAttitude[0];
	$xmB = @XmtrAttitude[1];
	$xmC = @XmtrAttitude[2];
	$xmD = @XmtrAttitude[3];
	
	$rcA = @RcvrAttitude[0];
	$rcB = @RcvrAttitude[1];
	$rcC = @RcvrAttitude[2];
	$rcD = @RcvrAttitude[3];
	
	$range = sqrt ($rX**2 + $rY**2 + $rZ**2);
	

	# this defines the return array
	my @returnArray = ();

	# RETURN YOUR RESULTS BELOW

	$returnArray[0] =  12.3e9;
	$returnArray[1] =  25.0e6;
	$returnArray[2] =  30;
	$returnArray[3] =  2.0;
	$returnArray[4] =  0.0;
	$returnArray[5] =  0;
	$returnArray[6] =  0.3;
	$returnArray[7] =  3;
	$returnArray[8] =  0;
	$returnArray[9] =  0.0;
	$returnArray[10] = 1.0;
	$returnArray[11] = -50.0;
	$returnArray[12] = 0.25;
	$returnArray[13] = 0.5;
	$returnArray[14] = 290.0;
	$returnArray[15] = 50.0;
	
	#############################################################################################
	# END OF USER MODEL AREA	
	#############################################################################################

	return @returnArray;
}


sub Perl_ReceiverModel_showArgs
{
	my @argStrArray;

	STKUtil::printOut "Doing Perl_ReceiverModel_compute_init\n";

	@argStrArray = ();

	push @argStrArray, $Perl_ReceiverModel_Inputs->{'FunctionName'}->{'Name'} . " Inputs \n";

	# the first arg on input is the calling mode

	push @argStrArray, "0 : this is the calling mode\n";

	my @args = $Perl_ReceiverModel_Inputs->getArgumentArray();

	# to see description args

	my $index, $descrip;

	foreach $arg (@args)
	{
		($index, $descrip) = $Perl_ReceiverModel_Inputs->getArgument($arg);

		push @argStrArray, "$index : $arg = $descrip\n";
	}

	STKUtil::printOut @argStrArray;

	@argStrArray = ();

	push @argStrArray, $Perl_ReceiverModel_Outputs->{'FunctionName'}->{'Name'} . " Outputs \n";

	my @args = $Perl_ReceiverModel_Outputs->getArgumentArray();

	# to see description args

	my $index, $descrip;

	foreach $arg (@args)
	{
		($index, $descrip) = $Perl_ReceiverModel_Outputs->getArgument($arg);

		push @argStrArray, "$index : $arg = $descrip\n";
	}

	STKUtil::printOut @argStrArray;

}

# MUST end Perl script file with a non-zero integer

1;